language/tools/move-bytecode-viewer/src/bytecode_viewer.rs

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+use crate::interfaces::{BytecodeInfo, LeftScreen};
+use bytecode_source_map::{mapping::SourceMapping, source_map::SourceMap};
+use disassembler::disassembler::{Disassembler, DisassemblerOptions};
+use move_ir_types::location::Loc;
+use regex::Regex;
+use std::collections::HashMap;
+use vm::{
+    access::ModuleAccess,
+    file_format::{CompiledModule, FunctionDefinitionIndex},
+};
+
+#[derive(Clone, Debug)]
+pub struct BytecodeViewer {
+    pub lines: Vec<String>,
+    pub module: CompiledModule,
+    pub line_map: HashMap<usize, BytecodeInfo>,
+}
+
+impl BytecodeViewer {
+    pub fn new(source_map: SourceMap<Loc>, module: CompiledModule) -> Self {
+        let source_mapping = SourceMapping::new(source_map, module.clone());
+        let mut options = DisassemblerOptions::default();
+        options.print_code = true;
+        options.print_basic_blocks = true;
+        let disassembled_string = Disassembler::new(source_mapping, options)
+            .disassemble()
+            .unwrap();
+
+        let mut base_viewer = Self {
+            lines: disassembled_string.lines().map(|x| x.to_string()).collect(),
+            line_map: HashMap::new(),
+            module,
+        };
+        base_viewer.build_mapping();
+        base_viewer
+    }
+
+    fn build_mapping(&mut self) {
+        let regex = Regex::new(r"^(\d+):.*").unwrap();
+        let fun_regex = Regex::new(r"^public\s+([a-zA-Z_]+)\(").unwrap();
```

Comment:
What is the goal of this (just to make sure I get it). Is it to get the position in the file of a certain function? 

We might want to do this by parsing the AST, I can help with that code if that is the goal

Reply:
You're correct. This is parsing the disassembled bytecode to build up a mapping of `line_number -> (function_def_index, code_offset)`.

 It would be best to create this mapping from the disassembler directly, but that will require updating the disassembler to a proper object model and then adding pretty printers to it (which is something we should do, but probably should be in a separate PR).

Reply:
I guess I'm confused why you don't just parse the file and run it a bit (up to expansion maybe) and grab the source location from the AST?

Reply:
We're grabbing the source location from the source map.

This is building up the mapping so that we know "when the cursor is at line 10, that corresponds to code offset 5 in function `foo`" so then we can grab the source map for that if that makes sense?